# Geoplot tutorial

## Geospatial data today

Geospatial data is any form of data which has a location to it. Most of the data generated today has such a
geospatial context, context which is, in turn, oftentimes important to understanding the data itself. Thanks to its
ease-of-use and deep ecosystem, the Python programming language has emerged as a leading choice in the performance of
data analytics, geospatial data analytics included. ``geoplot``, a tool for generating easy-to-use geospatial plot
types, builds on the fundamental peices of this existing programming stack, which we will discuss first.

The core abstraction is the ``GeoDataFrame``, an augmented version of a ``pandas`` ``DataFrame`` with an attached
geometry:

In [4]:
import geopandas as gpd
boroughs = gpd.read_file("./data/boroughs.geojson", driver='GeoJSON')
boroughs

BoroCode       BoroName    Shape_Area    Shape_Leng  \
0         5  Staten Island  1.623853e+09  330385.03697   
1         4         Queens  3.049947e+09  861038.47930   
2         3       Brooklyn  1.959432e+09  726568.94634   
3         1      Manhattan  6.364422e+08  358532.95642   
4         2          Bronx  1.186804e+09  464517.89055   

                                            geometry  
0  (POLYGON ((-74.05050806403247 40.5664220341608...  
1  (POLYGON ((-73.83668274106708 40.5949466970158...  
2  (POLYGON ((-73.8670614947212 40.58208797679338...  
3  (POLYGON ((-74.01092841268033 40.6844914725429...  
4  (POLYGON ((-73.89680883223775 40.7958084451597...

Readers familiar with ``pandas`` operations should be pleased to know that the library providing this
geometric abstraction, ``geopandas``, is just an extension of the former, and so has all the same fundamental
operations. That means that if you've ever worked with data before but haven't yet touched geospatial, the transition
should be completely straightforward. If you haven't worked with data before in Python and are thus unfamiliar with
``pandas``, you should refer to its documentation and get good with using that library first&mdash;[10 Minutes to Pandas](http://pandas.pydata.org/pandas-docs/stable/10min.html) is one place where you can get started.

Any operation that will work on a ``DataFrame`` will work on a ``GeoDataFrame``, but the latter adds a few additional
methods and features for dealing with geometry not present in the former. The most obvious of these is the addition
of a column for storing geometries, accessible using the ``geometry`` attribute:

In [5]:
boroughs.geometry

0    (POLYGON ((-74.05050806403247 40.5664220341608...
1    (POLYGON ((-73.83668274106708 40.5949466970158...
2    (POLYGON ((-73.8670614947212 40.58208797679338...
3    (POLYGON ((-74.01092841268033 40.6844914725429...
4    (POLYGON ((-73.89680883223775 40.7958084451597...
Name: geometry, dtype: object

That geometry is stored with reference to some kind of [**coordinate reference system**](https://en.wikipedia.org/wiki/Spatial_reference_system), or CRS. You can extract what CRS your polygons are stored in using the `crs` attribute:

In [6]:
boroughs.crs

{'init': 'epsg:4326'}

In this case `epsg:4326` is an identifier for what the rest of us more commonly refer to as "longitude and latitude". EPSG itself is a standardized system for refering to coordinate reference systems; [spatialreference.org](http://spatialreference.org/ref/epsg/wgs-84/) is the best place to look these identifiers up.

Coordinate reference systems are, basically, different ways of mathematically calculating locations. Due to the complexity of the surface of the earth, different geographically sensitive systems of measurement are more or less useful for different tasks. For example, the United States Geolocial Service, which provides extremely high-accuracy maps of United States localities, maintains individual coordinate reference systems, refered to as "state plane systems", for the various states of the union. These are used throughout government, and look nothing like the latitude and longitude coordinates that we are generally more used to.

For the purposes of illustration, here are a few coordinates of buildings in New York City, stored in [the original record](http://www1.nyc.gov/site/planning/data-maps/open-data.page) in the Long Island State Plane coordinate reference system:

Continue...

In [7]:
# TODO